%----------------------------------------------------------------------------------------
%	MODULE INFORMATION
%----------------------------------------------------------------------------------------

% Define the top matter
\setModuleTitle{Copy Number Variation}
\setModuleAuthors{%
  Velimir Gayevskiy \mailto{v.gayevskiy@garvan.org.au} \\
  Sonika Tyagi, AGRF \mailto{sonika.tyagi@agrf.org.au}%
}

\setModuleContributions{%
  XXX \mailto{xxx@xxx} \\
  yyy \mailto{yyy@yyy}% 
}

%----------------------------------------------------------------------------------------
%	MODULE TITLE PAGE
%----------------------------------------------------------------------------------------

\chapter{\moduleTitle}

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	LEARNING OUTCOMES
%----------------------------------------------------------------------------------------

\section{Key Learning Outcomes}

After completing this practical the trainee should be able to:

\begin{itemize}
  \item Understand and perform a simple copy number variation analysis on NGS data
  \item Become familiar with Sequenza and other CNV detection packages
  \item Understand the CNV inference process as an interplay between depth of sequencing, cellularity and B-allele frequency
  \item Visualize CNV events for manual inspection
\end{itemize}

%----------------------------------------------------------------------------------------
%	MODULE RESOURCES
%----------------------------------------------------------------------------------------

\section{Resources You'll be Using}

\subsection{Tools Used}

\begin{description}[style=multiline,labelindent=0cm,align=left,leftmargin=1cm]
  \item[Sequenza] \hfill\\
    \url{http://www.cbs.dtu.dk/biotools/sequenza/}
  \item[IGV] \hfill\\
    \url{http://www.broadinstitute.org/igv/}
\end{description}

%------------------------------------------------

\subsection{Sources of Data}

\url{http://sra.dnanexus.com/studies/ERP001071}\\
\url{http://www.ncbi.nlm.nih.gov/pubmed/22194472}

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	INTRODUCTION
%----------------------------------------------------------------------------------------

\section{Introduction}

The goal of this hands-on session is to perform a copy number variation analysis on a normal/tumour pair of alignment files (BAMs) produced by the mapping of Illumina short read sequencing. 

To ensure reasonable analysis times, we will perform the analysis on a heavily subset pair of BAM files. These files contain just the first 60Mb of chromosome 5 but contain several examples of inferred copy number events to enable interpretation and visualisation of the copy number variation that is present in entire cancer genomes. Sequenza is the tool we will use to perform this analysis. It consists of a Python pre-processing step followed by a second step in R to infer the depth ratio, cellularity, ploidy and to plot the results for interpretation.

In the second part of the tutorial we will also be using IGV to visualise and manually inspect the copy number variation we inferred in the first part for validation purposes. This section will also include a discussion on the importance of good quality data by highlighting the inadequacies of the workshop dataset and the implications this has on analysis results.

%----------------------------------------------------------------------------------------
%	THE ENVIRONMENT
%----------------------------------------------------------------------------------------

\section{Prepare the Environment}

We will use a dataset derived from whole genome sequencing of a 33-yr-old lung adenocarcinoma patient, who is a never-smoker and has no familial cancer history. 

The data files are contained in the subdirectory called \texttt{data} and are the following:

\begin{description}[style=multiline,labelindent=1.5cm,align=left,leftmargin=2.5cm]
  \item[\texttt{SM\_Blood.merged.mrkdup.realn.chr5.60Mb.bam} and \texttt{SM\_Blood.merged.mrkdup.realn.chr5.60Mb.bam.bai}] \hfill\\
  \item[\texttt{SM\_liverMets.merged.mrkdup.realn.chr5.60Mb.bam} and \texttt{SM\_liverMets.merged.mrkdup.realn.chr5.60Mb.bam}] \hfill\\
  These files are based on subsetting the whole genomes derived from blood and liver metastases to the first 60Mb of chromosome 5. This will allow our analyses to run in a sufficient time during the workshop, but it's worth being aware that we are analysing just 1.9\% of the genome which highlights the length of time and resources required to perform cancer genomics on full genomes!
\end{description}

\begin{steps}
Open the Terminal and go to the \texttt{CNV} working directory:
\begin{lstlisting}
cd ~/CNV/
\end{lstlisting}
\end{steps}

%\reversemarginpar\marginpar{\vskip+0em\hfill\includegraphics[height=1cm]{graphics/warning.png}}
%\textcolor{red}{
\begin{warning}
  All commands entered into the terminal for this tutorial should be from within the
  \textbf{\texttt{$\sim$/CNV}} directory.
\end{warning}

\begin{steps}
Check that the \texttt{data} directory contains the above-mentioned files by typing:
\begin{lstlisting}
ls data
\end{lstlisting}
\end{steps}

%----------------------------------------------------------------------------------------
%	SEQUENZA PRE-PROCESSING
%----------------------------------------------------------------------------------------

% pypy sequenza-utils.py bam2seqz -n "$1" -t "$2" --fasta genome.fa -gc sequenza.gc50Base.txt.gz -C $3 | gzip > sample_id_$3.large.seqz.gz
% pypy sequenza-utils.py seqz-binning -w 50 -s sample_id_$3.large.seqz.gz | tail -n +2 | gzip > sample_id_$3.seqz.gz
% parallel --gnu -v analyze "$normalbam_path" "$tumourbam_path" {} :::  X {1..22} Y

\section{Sequenza Pre-Processing}

Sequenza is run in three steps. The first pre-processing step is run on the final normal and tumour mapped data (BAM files) in order to walk the genome in a pileup format (automatically generated by samtools). This first step finds high quality sites in the genomes and extracts their depth and genotype in the normal genome and calculates the variant alleles and allele frequencies in the tumour genome. The second step is to perform a binning on these sites to save space and analysis time in the third step. Finally, the third step is run in R normalize the depth ratio between the normal/tumour genomes, infer cellularity and ploidy and graphically output results for interpretation.

\subsection{Step 1: Pre-Processing - Walking the Genome}

\begin{steps}
\begin{lstlisting}
pypy sequenza-utils.py bam2seqz -n data/SM\_Blood.merged.mrkdup.realn.chr5.60Mb.bam -t data/SM\_liverMets.merged.mrkdup.realn.chr5.60Mb.bam --fasta assets/genome_ref.fa -gc assets/sequenza.gc50Base.txt.gz | gzip > stage1-output.large.seqz.gz
\end{lstlisting}
\end{steps}

\begin{note}
Explanation of parameters
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
 \item[-n] the normal BAM
 \item[-t] the tumour BAM
 \item[--fasta] the reference genome used for mapping (b37 from GATK here)
 \item[-gc] GC content as windows through the genome (pre-generated and downloadable from the Sequenza website)
\end{description}
\end{note}

This will take approximately 30 minutes to run.

\begin{information}
You can look at the first few lines of the output in the file \texttt{stage1-output.large.seqz.gz} with:
 
\begin{lstlisting}
zcat stage1-output.large.seqz.gz | head -n 20
\end{lstlisting}
\end{information}

%------------------------------------------------

\subsection{Step 2: Perform Binning}

\begin{steps}
\begin{lstlisting}
pypy sequenza-utils.py seqz-binning -w 200 -s stage1-output.large.seqz.gz | gzip > stage1-output.binned.seqz.gz
\end{lstlisting}
\end{steps}

%\reversemarginpar\marginpar{\vskip+0em\hfill\includegraphics[height=1cm]{graphics/notes.png}}
\begin{note}
Explanation of parameters
\begin{description}[style=multiline,labelindent=0cm,align=right,leftmargin=\descriptionlabelspace,rightmargin=1.5cm,font=\ttfamily]
 \item[-w] the window size (50 for exomes, 200 for genomes)
 \item[-s] the large seqz file generated in the first step
\end{description}
\end{note}

This step should take approximately X minutes to complete.

%------------------------------------------------

\subsection{Step 3: Running Sequenza Algorithms and Plotting Results}

\begin{steps}
Open the R terminal
\begin{lstlisting}
R
\end{lstlisting}
\end{steps}

You should now see the R prompt identified with "> ".

\begin{steps}
Run the Sequenza R commands
\begin{lstlisting}
library("sequenza")
data.file <- "stage1-output.binned.seqz.gz"
seqzdata <- sequenza.extract(data.file)
CP.example <- sequenza.fit(seqzdata)
sequenza.results(sequenza.extract = seqzdata, cp.table = CP.example, sample.id = "CanGenWorkshop", out.dir="sequenza_results")
\end{lstlisting}
\end{steps}

If every command ran successfully, you will now have a "sequenza_results" folder containing 13 files.

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	SEQUENZA ANALYSIS VISUALISATION
%----------------------------------------------------------------------------------------

\section{Sequenza Analysis Results and Visualisation}


view PDFs

questions/interpratation

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	CNV VISUALISATION
%----------------------------------------------------------------------------------------

\section{CNV Visualisation/Confirmation in IGV}

Let's see if we can visualise one of the CNV events where copy number increased significantly. We'll focus on the copy number 4 event seen at about 1/3 of the way through the CanGenWorkshop_genome_view.pdf output. First, we need to find the coordinates that have been predicted for this event. Open the "CanGenWorkshop_segments.txt" file in the results folder to view all predicted CNV events. There is only one at a copy number of 4 and it starts at 21111248 to 21522065 which is 410kb and corresponds to the small block we see in the genome view PDF.

We will now open IGV and see if we can observe the predicted increase in copy number within these genomic coordinates.

\begin{steps}
Step by step instructions to open the BAMs in IGV and navigate to the coordinates of interest.

\begin{lstlisting}
command
\end{lstlisting}
\end{steps}

For a duplication of this size, we will not be able to easily observe it just by looking at the raw read alignments. In order to see it we will generate two tiled data files (TDFs) within IGV which contain average read depth for a given window size through the genome. This means that we can aggregate the average read depth over relatively large chunks of the genome and compare these values between the normal and tumour genomes.

To begin, we will go to "Tools" then "Run igvtools..." in the IGV menubar. Specify the normal bam file as the input file and change the window size to 100000 (one hundred thousand). Then press the "Run" button and IGV will make the TDF file. Repeat this for the tumour genome.

After you have both TDF files, go to "File" and "Load from file..." in the menubar and select both of them to open. Once you have opened them, they will appear as tracks along with the BAM tracks we loaded initially. Navigate back to the genomic coordinates of our event (5:21,111,248-21,522,065) and mouse over the two tracks to get the average depth values for the 100,000bp windows. What you should see is that the liverMets sample has 3-6 more coverage than the Blood sample for the four windows that cover this region.

This may seem a bit underwhelming, after all, wasn't the increase of the region to a copy number of 4, i.e. a doubling of reads in the tumour? To explain why we are only seeing such a small coverage increase, we need to turn to our good friend math.

Imagine we have two 30X genomes for the normal and tumour samples and the tumour is at 100\% purity. If there is a copy number increase to 4 in the tumour from 2 in the normal, the duplicated segment should indeed have twice as many reads as the same segment in the normal genome. Now, lets imagine the tumour genome was only at a purity of 50\% (i.e. it contains 50\% normal cells and 50\% tumour cells). Now, half of the duplicated "tumour genome" segment will be at a copy number of 2 and half will be at 4. What does this mean when we sequence them as a mixture? The resulting average read depth increase of the block will be $(0.5*2)+(0.5*4) = 3$. Now what if we only have 20\% tumour cells in our "tumour genome"? This will be $(0.8*2)+(0.2*4) = 2.4$. You can see how sequencing a low cellularity tumour at a low depth is makes it much harder to infer copy number variations!

Returning to our genomes at hand, since we saw a copy number increase of just 3-6X (about 10-20\%), perhaps our tumour genome is at a low cellularity? To find out, lets look at the Sequenza estimate of cellularity which we can find in the "CanGenWorkshop_model_fit.pdf" file in the results folder. Indeed, it looks like Sequenza has inferred the cellularity to be just 0.24 indicating that the "tumour genome" actually has quite a lot of normal cell contamination which reduces our power to infer copy numbers.

\begin{note}
It is possible to sequence through a low-cellularity sample when, for example, there is no way to take another sample as in the case of most biopsies. Sequencing through means to simply sequence the tumour at a much higher coverage, usually 90-120X. This will mean that there will be a total increase in reads supplying evidence for copy number events and variants and in aggregate these will still retain power to infer these events when using tools that look at the whole genome like Sequenza does.
\end{note}

%----------------------------------------------------------------------------------------

\newpage

%----------------------------------------------------------------------------------------
%	REFERENCES
%----------------------------------------------------------------------------------------

\section{References}

%TODO Change to using BiBTeX
\begin{enumerate}
  \item example only: Trapnell, C., Pachter, L. \& Salzberg, S. L. TopHat: discovering splice
  junctions with RNA-Seq. Bioinformatics 25, 1105-1111 (2009).
  
\end{enumerate}
